util/collate/unicode_ci.go

```diff
@@ -36,12 +160,133 @@ type unicodePattern struct {
 	patTypes []byte
 }
 
-// Compile implements WildcardPattern interface. Do nothing temporary, will change when implement
+// Compile implements WildcardPattern interface.
 func (p *unicodePattern) Compile(patternStr string, escape byte) {
-
+	p.patChars, p.patTypes = compilePatternUnicodeCI(patternStr, escape)
 }
 
-// DoMatch implements WildcardPattern interface. Always return false temporary, will change when implement
+// DoMatch implements WildcardPattern interface.
 func (p *unicodePattern) DoMatch(str string) bool {
-	return false
+	return doMatchUnicodeCI(str, p.patChars, p.patTypes)
+}
+
+// compilePatternUnicodeCI handles escapes and wild cards, generate pattern weights and types.
+// This function is modified from stringutil.CompilePattern.
+func compilePatternUnicodeCI(pattern string, escape byte) (patWeights []rune, patTypes []byte) {
+	runes := []rune(pattern)
+	escapeRune := rune(escape)
+	lenRunes := len(runes)
+	patWeights = make([]rune, lenRunes)
+	patTypes = make([]byte, lenRunes)
+	patLen := 0
+	for i := 0; i < lenRunes; i++ {
+		var tp byte
+		var r = runes[i]
+		switch r {
+		case escapeRune:
+			tp = stringutil.PatMatch
+			if i < lenRunes-1 {
+				i++
+				r = runes[i]
+				if r == escapeRune || r == '_' || r == '%' {
+					// Valid escape.
+				} else {
+					// Invalid escape, fall back to escape byte.
+					// mysql will treat escape character as the origin value even
+					// the escape sequence is invalid in Go or C.
+					// e.g., \m is invalid in Go, but in MySQL we will get "m" for select '\m'.
+					// Following case is correct just for escape \, not for others like +.
+					// TODO: Add more checks for other escapes.
+					i--
+					r = escapeRune
+				}
+			}
+		case '_':
+			// %_ => _%
+			if patLen > 0 && patTypes[patLen-1] == stringutil.PatAny {
+				tp = stringutil.PatAny
+				r = '%'
+				patWeights[patLen-1], patTypes[patLen-1] = '_', stringutil.PatOne
+			} else {
+				tp = stringutil.PatOne
+			}
+		case '%':
+			// %% => %
+			if patLen > 0 && patTypes[patLen-1] == stringutil.PatAny {
+				continue
+			}
+			tp = stringutil.PatAny
+		default:
+			tp = stringutil.PatMatch
+		}
+		patWeights[patLen] = r
+		patTypes[patLen] = tp
+		patLen++
+	}
+	patWeights = patWeights[:patLen]
+	patTypes = patTypes[:patLen]
+	return
+}
+
+// doMatchUnicodeCI matches the string with patWeights and patTypes.
+// The algorithm has linear time complexity.
+// https://research.swtch.com/glob
+// This function is modified from stringutil.DoMatch.
+func doMatchUnicodeCI(str string, patWeights []rune, patTypes []byte) bool {
```

Comment:
Is it the same as doMatchGeneralCI ?

Reply:
almost same, but is different in detial, we may unify this, include `stringutil.DoMatch` in another PR